staging/src/k8s.io/apiserver/pkg/endpoints/filters/authn_audit.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/audit/policy"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// WithFailedAuthenticationAudit decorates a http.Handler with a fallback audit,
+// logging information only when the original one did was not triggered.
+// This needs to be used with WithAuditTriggeredMarker, which wraps the original
+// audit filter.
+func WithFailedAuthenticationAudit(handler http.Handler, requestContextMapper request.RequestContextMapper, sink audit.Sink, policy policy.Checker) http.Handler {
+	if sink == nil || policy == nil {
+		return handler
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx, ok := requestContextMapper.Get(req)
+		if !ok {
+			responsewriters.InternalError(w, req, errors.New("no context found for request"))
+			return
+		}
+
+		ev, err := createAuditEvent(ctx, policy, requestContextMapper, req, w)
+		if err != nil {
+			return
+		}
+		if ev == nil {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		ev.ResponseStatus = &metav1.Status{}
+		ev.ResponseStatus.Message = getAuthMethods(req)
+		ev.Stage = auditinternal.StageResponseStarted
+
+		rw := decorateResponseWriter(w, ev, sink)
```

Comment:
This makes the rw.WriteHeader to process an event. The `WithAudit` also decorates. So if an endpoint handler writes the header, the `WithAudit` decorator will send an event to the audit backend, then your new outer decorator will the same. Moreover, you `ev` here is another one than the one of `WithAudit`. So yours will be pretty empty.

Reply:
No double decoration will happen. he reason for that is that `WithFailedAuthenticationAudit` decorates *only* the failed handler that is passed to `WithAuthentication`. At that point no request will ever reach regular `WithAudit` because the processing is passed through a different handler chain.

Reply:
got it. Please call the parameter `failedHandler` here. Then it's clearer.